File: GeoGuessSwipe/app/src/main/java/com/example/dhrdjordjevic/geoguess/MainActivity.java
package com.example.dhrdjordjevic.geoguess;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;


public class MainActivity extends AppCompatActivity {

    GeoObjectAdapter mAdapter;
    RecyclerView mGeoRecyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final List<GeoObject> mGeoObjects = new ArrayList<>();

        for(int i = 0; i < GeoObject.PRE_DEFINED_GEO_OBJECT_NAMES.length; i++)
        {
            mGeoObjects.add(new GeoObject(GeoObject.PRE_DEFINED_GEO_OBJECT_NAMES[i],
                    GeoObject.PRE_DEFINED_GEO_OBJECT_IMAGE_IDS[i]));
        }

        mGeoRecyclerView = findViewById(R.id.recyclerView);

        RecyclerView.LayoutManager mLayoutManager = new
                StaggeredGridLayoutManager(1, LinearLayoutManager.VERTICAL);

        mGeoRecyclerView.setLayoutManager(mLayoutManager);
        mAdapter = new GeoObjectAdapter(this,
                mGeoObjects);
        mGeoRecyclerView.setAdapter(mAdapter);

        ItemTouchHelper.SimpleCallback simpleItemTouchCallback =
                new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT)
                {
                    @Override
                 public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder
                            target)
                    {
                        return false;
                    }

                    //Called when a user swipes left or right on a ViewHolder
                    @Override

                 public void onSwiped(RecyclerView.ViewHolder viewHolder, int swipeDir)
                {
                    //Get the index corresponding to the selected position
                    int position = (viewHolder.getAdapterPosition());

                    if(mAdapter.isInEurope(position).equals("yes") && swipeDir == ItemTouchHelper.LEFT)
                    {
                        Toast.makeText(getApplicationContext(),"Good job!", Toast.LENGTH_SHORT).show();
                    } else if(mAdapter.isInEurope(position).equals("no") && swipeDir == ItemTouchHelper.RIGHT)
                    {
                        Toast.makeText(getApplicationContext(),"Good job!", Toast.LENGTH_SHORT).show();
                    } else
                    {
                        Toast.makeText(getApplicationContext(),"You Suck!", Toast.LENGTH_SHORT).show();
                    }

                    mGeoObjects.remove(position);
                    mAdapter.notifyItemRemoved(position);
                }
                };

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);
        itemTouchHelper.attachToRecyclerView(mGeoRecyclerView);
    }
}

File: kotlinSwipeQuiz/app/src/main/java/com/example/swipequiz/MainActivity.kt
package com.example.swipequiz

import android.content.ClipData
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.contentValuesOf
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_main.*
import java.nio.file.attribute.PosixFileAttributeView

class MainActivity : AppCompatActivity() {

    private val questions = arrayListOf<Question>()

    //setting the listener to a method to show answer
    private val questionAdapter = QuestionAdapter(questions, {question : Question -> showAnswer(question)})

    private lateinit var constraintLayout: ConstraintLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        constraintLayout = findViewById(R.id.constraintLayout)
        initViews()
    }

    private fun initViews()
    {
        rvQuestions.layoutManager = LinearLayoutManager(this@MainActivity, RecyclerView.VERTICAL, false)
        rvQuestions.adapter = questionAdapter
        rvQuestions.addItemDecoration(DividerItemDecoration(this@MainActivity, DividerItemDecoration.VERTICAL))

        for (i in Question.QUESTIONS.indices)
        {
            questions.add(Question(Question.QUESTIONS[i], Question.ANSWERS[i]))
        }

        createItemTouchHelper().attachToRecyclerView(rvQuestions)

        questionAdapter.notifyDataSetChanged()
    }

    private fun createItemTouchHelper(): ItemTouchHelper
    {
        val callback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT)
        {
            override fun onMove(
                recyclerView: RecyclerView,
                viewHolder: RecyclerView.ViewHolder,
                target: RecyclerView.ViewHolder
                ): Boolean
            {
                return true
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.adapterPosition

                if(direction == ItemTouchHelper.LEFT) onSwipeLeft(position)
                else if(direction == ItemTouchHelper.RIGHT) onSwipeRight(position)

                questionAdapter.notifyItemChanged(viewHolder.adapterPosition)
            }


        }
        return ItemTouchHelper(callback)
    }

    private fun onSwipeLeft(pos: Int)
    {
        if(questions[pos].answer) {
            val snack = Snackbar.make(constraintLayout, R.string.correct, Snackbar.LENGTH_LONG)
            snack.show()
        }
        else
        {
            val  snack = Snackbar.make(constraintLayout, R.string.wrong, Snackbar.LENGTH_LONG)
            snack.show()
        }
    }

    private fun onSwipeRight(pos: Int)
    {
        if(!questions[pos].answer) {
            val snack = Snackbar.make(constraintLayout, R.string.correct, Snackbar.LENGTH_LONG)
            snack.show()
        }
        else
        {
            val snack = Snackbar.make(constraintLayout, R.string.wrong, Snackbar.LENGTH_LONG)
            snack.show()
        }
    }

    //method to use in the viewholder listener
    private fun showAnswer(question: Question)
    {
        val snack = Snackbar.make(constraintLayout, "This statement is " + question.answer + "!", Snackbar.LENGTH_LONG)
        snack.show()
    }
}
